runtime/src/main/java/org/corfudb/runtime/collections/StreamSubscription.java

```diff
@@ -127,8 +132,19 @@ boolean enqueueStreamEntry(ILogData logData, long maxBlockTime) throws Interrupt
                     .stream()
                     .map(entry -> CorfuStreamEntry.fromSMREntry(entry, epoch))
                     .collect(Collectors.toList());
+
+            // Deduplicate entries per stream Id
+            Map<Message, CorfuStreamEntry> observedKeys = new LinkedHashMap<>();
+            entryList.forEach(entry -> {
+                if (observedKeys.putIfAbsent(entry.getKey(), entry) != null) {
+                    // If the key is present, remove and insert again to preserve transaction order
+                    observedKeys.remove(entry.getKey());
+                    observedKeys.put(entry.getKey(), entry);
+                }
```

Comment:
Looks like we can simply call `put(entry.getKey(), entry)` to deduplicate

Reply:
the problem is that we want to preserve the order, so the case: 
k1, v1
k1, v2
k2, v3
k1, v4 -> if we don't remove you'll report the updates in the following way:

k1, v4 -> k2, v3 (which is incorrect).. and I tried the replace function but that preserves the initial insertion order

Reply:
Why do we need to preserve order between unique keys? 

Maybe it should be documented on the subscription API, if its necessary. It also seems like challenging to code against for such semantics. That means that the consumer needs to encode some relationship between the unique keys for this order to matter.

Reply:
I do see a good point here about why we need to preserve the order across unique keys within a transaction. Current expectation is to emit the final version of a key within the transaction but does not talk about any order of updates within a transaction. Please share your thoughts on how this can be beneficial for the clients?

Reply:
Consider the following operation in the same transaction:
enqueue(A)
enqueue(A`)
if we deliver A after A` then the apply will result in incorrect version of the object right?